jbmc/src/java_bytecode/assignments_from_json.cpp

```diff
@@ -0,0 +1,828 @@
+{
+  const pointer_typet &pointer_type = to_pointer_type(expr.type());
+  const auto &element_type =
+    static_cast<const typet &>(pointer_type.subtype().find(ID_element_type));
+  side_effect_exprt java_new_array{ID_java_new_array, pointer_type};
+  java_new_array.copy_to_operands(array_length_expr);
+  java_new_array.type().subtype().set(ID_element_type, element_type);
+  code_assignt assign{expr, java_new_array, info.loc};
+  info.block.add(assign);
+}
+
+/// One of the cases in the recursive algorithm: the case where \p expr
+/// represents an array.
+/// The length of the array is given by a symbol: \p given_length_expr if it is
+/// specified (this will be the case when there are two or more reference-equal
+/// arrays), or a fresh symbol otherwise.
+/// If \p given_length_expr is specified, we assume that an array with this
+/// symbol as its length has already been allocated and that \p expr has been
+/// assigned to it.
+/// Either way, the length symbol stores a nondet integer, and we add
+/// constraints on this: if "nondetLength" is specified in \p json, then the
+/// number of elements specified in \p json should be the minimum length of the
+/// array. Otherwise the number of elements should be the exact length of the
+/// array.
+/// For the assignment of the array elements, see
+/// \ref assign_array_data_component_from_json.
+/// For the overall algorithm, see \ref assign_from_json_rec.
+static void assign_array_from_json(
+  const exprt &expr,
+  const jsont &json,
+  const optionalt<symbol_exprt> &given_length_expr,
+  const optionalt<std::string> &type_from_array,
+  det_creation_infot &info)
+{
+  PRECONDITION(has_array_type(expr, info.symbol_table));
+  PRECONDITION(can_cast_type<pointer_typet>(expr.type()));
+  const auto &element_type = static_cast<const typet &>(
+    to_pointer_type(expr.type()).subtype().find(ID_element_type));
+
+  exprt length_expr;
+  if(given_length_expr)
+    length_expr = *given_length_expr;
+  else
+  {
+    length_expr = info.allocate_objects.allocate_automatic_local_object(
+      java_int_type(), "det_array_length");
+    info.block.add(code_assignt{
+      length_expr, side_effect_expr_nondett{java_int_type(), info.loc}});
+    info.block.add(code_assumet{binary_predicate_exprt{
+      length_expr, ID_ge, from_integer(0, java_int_type())}});
+    allocate_array(expr, length_expr, info);
+  }
+  const json_arrayt json_array = get_untyped_array(json, element_type);
+  const auto number_of_elements =
+    from_integer(json_array.size(), java_int_type());
+  info.block.add(code_assumet{
+    binary_predicate_exprt{length_expr, ID_ge, number_of_elements}});
+  if(has_nondet_length(json))
+  {
+    info.block.add(code_assumet{binary_predicate_exprt{
+      length_expr,
+      ID_le,
+      from_integer(info.max_user_array_length, java_int_type())}});
+  }
+  else
+  {
+    info.block.add(code_assumet{
+      binary_predicate_exprt{length_expr, ID_le, number_of_elements}});
```

Comment:
If I understand correctly, in the deterministic case, we add these two statements:
`length_expr <= number_of_elements`
`length_expr >= number_of_elements`
Why add an equality assumption in that case instead?

Reply:
That was something I considered - doing it this way seemed easier to do since one of the constraints is always necessary, and I'm not sure if an equality constraint would really simplify things for the solver?

Reply:
From what I understand about CDCLs, an equality expression should be better as it allows boolean constraint propagation. An equality is also easier to read when debugging the VCCs, and it wouldn't make the present code any more complicated. I suggest a follow up PR for that.